metecho/api/serializers.py

```diff
@@ -398,21 +385,60 @@ def get_pr_url(self, obj) -> Optional[str]:
-        return super().create(validated_data)
 
-    def update(self, instance, validated_data):
-        user = getattr(self.context.get("request"), "user", None)
-        originating_user_id = str(user.id) if user else None
-        self._handle_reassign(
-            "dev", instance, validated_data, user, originating_user_id
-        )
-        self._handle_reassign("qa", instance, validated_data, user, originating_user_id)
-        validated_data.pop("should_alert_dev", None)
-        validated_data.pop("should_alert_qa", None)
-        return super().update(instance, validated_data)
+class TaskAssigneeSerializer(serializers.Serializer):
+    assigned_dev = serializers.CharField(allow_null=True, required=False)
+    assigned_qa = serializers.CharField(allow_null=True, required=False)
+    should_alert_dev = serializers.BooleanField(required=False)
+    should_alert_qa = serializers.BooleanField(required=False)
+
+    def validate(self, data):
+        if "assigned_qa" not in data and "assigned_dev" not in data:
+            raise serializers.ValidationError(
+                "You must assign a developer, tester, or both"
+            )
+        return super().validate(data)
+
+    def validate_assigned_dev(self, new_dev):
+        user = self.context["request"].user
+        task: Task = self.instance
+        if not task.has_push_permission(user):
+            raise serializers.ValidationError(
+                "You don't have permissions to change the assigned developer"
+            )
+        if new_dev and not task.epic.project.is_collaborator(new_dev):
+            raise serializers.ValidationError(
+                f"User is not a valid GitHub collaborator: {new_dev}"
+            )
+        return new_dev
+
+    def validate_assigned_qa(self, new_qa):
+        user = self.context["request"].user
+        task: Task = self.instance
+        if not task.has_push_permission(user):
+            is_removing_self = new_qa is None and task.assigned_qa == user.github_id
+            is_assigning_self = new_qa == user.github_id and task.assigned_qa is None
+            if not (is_removing_self or is_assigning_self):
+                raise serializers.ValidationError(
+                    "You can only assign/remove yourself as a tester"
+                )
+        if new_qa and not task.epic.project.is_collaborator(new_qa):
+            raise serializers.ValidationError(
+                f"User is not valid GitHub collaborator: {new_qa}"
+            )
+        return new_qa
+
+    def update(self, task, data):
+        user = self.context["request"].user
+        user_id = str(user.id)
+        if "assigned_dev" in data:
+            task.assigned_dev = data["assigned_dev"]
+            self._handle_reassign("dev", task, data, user, originating_user_id=user_id)
+        if "assigned_qa" in data:
+            task.assigned_qa = data["assigned_qa"]
+            self._handle_reassign("qa", task, data, user, originating_user_id=user_id)
```

Comment:
@jerivas I think the order needs to be switched on these, so that in `_handle_reassign` we can compare the existing and new assignees before making the change.

```suggestion
        if "assigned_dev" in data:
            self._handle_reassign("dev", task, data, user, originating_user_id=user_id)
            task.assigned_dev = data["assigned_dev"]
        if "assigned_qa" in data:
            self._handle_reassign("qa", task, data, user, originating_user_id=user_id)
            task.assigned_qa = data["assigned_qa"]
```

Reply:
Fixed in https://github.com/oddbird/Metecho/commit/cadddfee10f9d33a1f554afeee8264ba3bb1b020

Reply:
Thanks, you hit this while writing the tests as well